src/gui/http.go

```diff
@@ -74,39 +89,40 @@ func LaunchWebInterfaceHTTPS(host, staticDir string, daemon *daemon.Daemon, cert
 		return err
 	}
 
-	listener, err = tls.Listen("tcp", host, &tls.Config{Certificates: certs})
+	s.listener, err = tls.Listen("tcp", host, &tls.Config{Certificates: certs})
 	if err != nil {
 		return err
 	}
 
 	// Runs http.Serve() in a goroutine
-	serve(listener, NewGUIMux(appLoc, daemon), quit)
 	return nil
+
 }
 
-func serve(listener net.Listener, mux *http.ServeMux, q chan struct{}) {
-	go func() {
-		for {
-			if err := http.Serve(listener, mux); err != nil {
-				select {
-				case <-q:
-					return
-				default:
-				}
-				continue
+func (s *Server) Serve() error {
+	defer logger.Info("Server closed")
+	logger.Info("Starting web interface on %s", s.listener.Addr())
+	mux := NewGUIMux(s.appLoc, s.daemon)
+	for {
+		if err := http.Serve(s.listener, mux); err != nil {
+			select {
+			case <-s.done:
+				return nil
+			default:
+				return err
 			}
+			continue
 		}
-	}()
+	}
 }
 
 // Shutdown close http service
-func Shutdown() {
-	if quit != nil {
-		// must close quit first
-		close(quit)
-		listener.Close()
-		listener = nil
-	}
+func (s *Server) Shutdown() {
+	logger.Info("Shutting down Server")
+	// must close done first
+	close(s.done)
```

Comment:
You misunderstood, the `Serve()` loop closes a `done` channel. At the end of `Shutdown`, wait for it, i.e. `<-s.done`